### ISSUE TITLE
[Troubleshoot]: Unexpected performance degradation after SFT fine-tuning Pi0

### ISSUE THREAD

**Issue by red0orange** [RLinf/RLinf#540]
### Problem description

I fine-tuned **Pi0 with SFT** using the official checkpoint **`RLinf-Pi0-LIBERO-Spatial-Object-Goal-SFT`** together with the official **LIBERO dataset** (**`physical-intelligence/libero`**) for **200 SFT steps**. However, I observed an **unexpected performance regression** after training.

The training run **did not report any errors**, and the logs/curves look normal overall. The dataset source is also consistent with what the authors mentioned in this issue: [https://github.com/RLinf/RLinf/issues/494](https://github.com/RLinf/RLinf/issues/494).

Could you please help identify what might be causing the degradation (e.g., config mismatch, evaluation protocol differences, normalization/tokenizer issues, or other hidden pitfalls)? Any guidance would be appreciated.

<img width="1430" height="1114" alt="Image" src="https://github.com/user-attachments/assets/0b137ac1-694c-4489-9b14-fad1d01580b3" />

### Configuration YAML file
Train：

[my_libero_sft_openpi.yaml](https://github.com/user-attachments/files/24488306/my_libero_sft_openpi.yaml)

Eval：

[my_libero_goal_ppo_openpi_eval.yaml](https://github.com/user-attachments/files/24488308/my_libero_goal_ppo_openpi_eval.yaml)

### Log file

[run_embodiment.log](https://github.com/user-attachments/files/24488188/run_embodiment.log)

### Environment
RLinf version: latest

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by red0orange** [comment]
I used the same model and dataset with the official OpenPi fine-tuning implementation and obtained the results below, which appear to be normal. I am curious about the cause of this discrepancy.

<img width="2571" height="1410" alt="Image" src="https://github.com/user-attachments/assets/e52a1369-5f11-4131-9293-09e62d8b2409" />

**Comment by red0orange** [comment]
<img width="3697" height="547" alt="Image" src="https://github.com/user-attachments/assets/0859d920-184e-40ec-835c-b7cf77b04f5c" />

With the help of developer @XuS1994, I am now able to fully reproduce the official OpenPI training curves. I made three changes in total; however, one of them involves a very strange bug where I need your help:
1. **Switched to full-parameter fine-tuning:** I changed `train_expert_only` to `False`.
2. **Aligned the optimizer:** I completely replicated the optimizer parameters and strategy from the official OpenPI configuration (this only required changes to the config file).
3. **The strange bug:** With the two changes above, single-GPU training works perfectly. However, for multi-GPU training, I found a bug in `fsdp_sft_worker` regarding the automatic gradient synchronization (all-reduce). After `loss.backward()`, I verified that there is a blocking wait for synchronization, but the gradients are not actually syncing correctly. This results in different `grad_norm` values on different cards, which is very odd. I confirmed this is a bug (at least in my current environment). If I manually add a full-parameter gradient synchronization step after the backward pass, everything works fine. While this unblocks me, the underlying bug in the worker still needs to be resolved, and I currently have no leads on the root cause.

config file: 
[my_libero_sft_openpi.yaml](https://github.com/user-attachments/files/24556679/my_libero_sft_openpi.yaml)

gradient synchronization after `loss.backward()`:
```python
...
    self.grad_scaler.scale(loss).backward()  

################# Grad Sync Bug Fix #################
torch.cuda.synchronize()
if torch.distributed.is_initialized():
    all_grads = []
    for name, param in self.model.named_parameters():
        if param.grad is not None:
            all_grads.append((name, param.grad))
        process_group = getattr(self.model, 'process_group', None)
        for name, grad in all_grads:
            torch.distributed.all_reduce(grad, op=torch.distributed.ReduceOp.AVG, group=process_group)
        # Final barrier to ensure all gradient synchronization is complete
        process_group = getattr(self.model, 'process_group', None)
        torch.distributed.barrier(group=process_group, device_ids=[torch.cuda.current_device()])
################# Grad Sync Bug Fix #################
...
```

**Comment by red0orange** [comment]
This issue has been fixed by the PR #554.

**Comment by AaronCaoZJ** [comment]
@red0orange Thank you for sharing your experimental configuration and experience. I have a question: why does SFT get reward, as seen in your table, as in my mind rollout and reward only appear in RL?

**Comment by red0orange** [comment]
> [@red0orange](https://github.com/red0orange) Thank you for sharing your experimental configuration and experience. I have a question: why does SFT get reward, as seen in your table, as in my mind rollout and reward only appear in RL?

Actually, I used the weights from the SFT training and ran them through the official evaluation script. These are the evaluation results.
